server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetMappingAction.java

```diff
@@ -106,73 +107,95 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
+                        new RestGetMappingsResponse(types, getMappingsResponse, threadPool::relativeTimeInMillis, timeout)));
+    }
+
+    private static final class RestGetMappingsResponse implements StatusToXContentObject {
+        private final String[] types;
+        private final GetMappingsResponse response;
+        private final LongSupplier relativeTimeSupplierMillis;
+        private final TimeValue timeout;
+        private final long startTimeMs;
+        private final SortedSet<String> missingTypes;
+
+        private RestGetMappingsResponse(String[] types,
+                                        GetMappingsResponse response,
+                                        LongSupplier relativeTimeSupplierMillis,
+                                        TimeValue timeout) {
+            this.types = types;
+            this.response = response;
+            this.relativeTimeSupplierMillis = relativeTimeSupplierMillis;
+            this.timeout = timeout;
+            this.startTimeMs = relativeTimeSupplierMillis.getAsLong();
+            this.missingTypes = Collections.unmodifiableSortedSet(getMissingTypes(types, response.getMappings()));
+        }
+
+        @Override
+        public RestStatus status() {
+            return missingTypes.isEmpty() ? RestStatus.OK : RestStatus.NOT_FOUND;
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            if (relativeTimeSupplierMillis.getAsLong() - startTimeMs > timeout.millis()) {
+                throw new ElasticsearchTimeoutException("Timed out getting mappings");
+            }
+
+            final ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetadata>> mappingsByIndex = response.getMappings();
+            if (mappingsByIndex.isEmpty() && types.length != 0) {
+                builder.close();
+                throw new TypeMissingException("_all", String.join(",", types));
             }
-        });
+
+            builder.startObject();
+            {
+                if (missingTypes.isEmpty() == false) {
+                    final String message = String.format(Locale.ROOT, "type" + (missingTypes.size() == 1 ? "" : "s") +
+                        " [%s] missing", Strings.collectionToCommaDelimitedString(missingTypes));
+                    builder.field("error", message);
+                    builder.field("status", status().getStatus());
+                }
+                response.toXContent(builder, params);
+            }
+            builder.endObject();
+
+            return builder;
+        }
+
+        private static SortedSet<String> getMissingTypes(String[] types,
+                                                   ImmutableOpenMap<String, ImmutableOpenMap<String, MappingMetadata>> mappingsByIndex) {
+            if (mappingsByIndex.isEmpty() && types.length != 0) {
+                return Arrays.stream(types).collect(Collectors.toCollection(TreeSet::new));
```

Comment:
In this case we are going to throw a `TypeMissingException`, so I don't understand why we do work iterating over `types` here. Can we just return an empty set?

Reply:
I did it for consistency, i.e. if we couldn't find any of the requested types, all of them are missing. But I agree that's a bit wasteful, I'll change it. 👍